common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
-        try {
-            boolean foundFromCache = true;
-            List<Class> extensions = providers.get(service);
-            if (extensions == null) {
-                synchronized (service) {
-                    extensions = providers.get(service);
-                    if (extensions == null) {
-                        extensions = findAllExtensionClass(service, activateName, loader);
-                        foundFromCache = false;
-                        providers.put(service, extensions);
-                    }
+    static class InnerEnhancedServiceLoader<S> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(EnhancedServiceLoader.class);
+        private static final String SERVICES_DIRECTORY = "META-INF/services/";
+        private static final String SEATA_DIRECTORY = "META-INF/seata/";
+
+        private static final ConcurrentMap<Class<?>, InnerEnhancedServiceLoader<?>> SERVICE_LOADERS =
+                new ConcurrentHashMap<>();
+        private static final ConcurrentMap<Class<?>, Object> EXTENSION_INSTANCES = new ConcurrentHashMap<>();
+
+        private final Class<S> type;
+        private final Holder<Map<ExtensionURL, Class<?>>> extensionClasses = new Holder<>();
+        private final ConcurrentMap<ExtensionURL, Holder<Object>> extensionInstances = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, List<ExtensionURL>> extensionNameUrlsMap = new ConcurrentHashMap<>();
+        private final ConcurrentMap<Class<?>, ExtensionURL> extensionClassUrlMap = new ConcurrentHashMap<>();
+        private ExtensionURL defaultExtensionURL = null;
+        private Integer highestLoadPriority = Integer.MIN_VALUE;
+
+        private InnerEnhancedServiceLoader(Class<S> type) {
+            this.type = type;
+        }
+
+        /**
+         * Get the ServiceLoader for the specified Class
+         *
+         * @param type the type of the extension point
+         * @param <S>  the type
+         * @return the service loader
+         */
+        static <S> InnerEnhancedServiceLoader<S> getServiceLoader(Class<S> type) {
+            if (type == null) {
+                throw new IllegalArgumentException("Enhanced Service type == null");
+            }
+            InnerEnhancedServiceLoader<S> loader = (InnerEnhancedServiceLoader<S>)SERVICE_LOADERS.get(type);
+            if (loader == null) {
+                SERVICE_LOADERS.putIfAbsent(type, new InnerEnhancedServiceLoader<S>(type));
+                loader = (InnerEnhancedServiceLoader<S>)SERVICE_LOADERS.get(type);
+            }
+            return loader;
+        }
+
+        /**
+         * Specify classLoader to load the service provider
+         *
+         * @param loader the loader
+         * @return s s
+         * @throws EnhancedServiceNotFoundException the enhanced service not found exception
+         */
+        public S load(ClassLoader loader) throws EnhancedServiceNotFoundException {
```

Comment:
The access modifiers here should be private.

Reply:
changed